src/libtriton/engines/symbolic/symbolicExpression.cpp

```diff
@@ -155,11 +155,27 @@ namespace triton {
 
 
       void SymbolicExpression::setAst(const triton::ast::SharedAbstractNode& node) {
-        for(auto sp : this->ast->getParents()) {
-          node->setParent(sp.get());
+        auto old = this->ast;
+        if (node == old)
+          return;
+
+        if (old) {
+          for (auto sp : old->getParents()) {
+            node->setParent(sp.get());
+          }
         }
+
         this->ast = node;
-        this->ast->init();
+
+        if ( !old
+          || (old->isSymbolized() != ast->isSymbolized())
+          || (old->evaluate() != ast->evaluate())
+          || (old->getBitvectorSize() != ast->getBitvectorSize())
+          || (old->isLogical() != ast->isLogical())
+          ) {
+          this->ast->initParents();
+        }
+
```

Comment:
I think it's more clear to let `this->ast->init();` For example, if in the future i'm adding new attributes in an AbstractNode, i have to remember to also add this attribute in this condition. So, I prefer to keep `ast->init()` :).

Reply:
1/ why init and not initParents ?
2/ I agree it is clearly not the right place for this design-wise. However it is important when converting expressions to symbolic variables if they do already have huge parent trees. I added this when a scenario spent ~95% of the time in there.
I just added AstNode::hasSameConcreteValueAndType in in my next push. It would be interesting to use it in Simplification passes too to check that the given callbacks do give back proper nodes. When using multiple simplification callbacks, if one does change the value type it would only detected after the end of the processing loop when its new parent node initialize.
It can also be left undetected until late if the node is the one currently being assigned to an expression..

Reply:
Got it but I still have to brainstorm about it. What I propose: Split this MR in two MRs. One without this feature so that I can merge all previous fixes and another one with this feature so that we can still discus about it and find a good way to handle this.

Reply:
It's ok, i don't care about the history, you can rebase and force push.